docs/rfc/0004_component_groups_scale_view.md

```diff
@@ -0,0 +1,299 @@
+four independent components, and each of the component can have one or more groups. Therefore, the `kubectl scale` can
+not work directly on the `RisingWave` resource and so do other tools depend on that. And it is the main reason for
+proposing this new CR. I named it 'RisingWaveComponentScaleView'.
+
+> What use cases does it support?
+
+With the `RisingWaveComponentScaleView`, the users can set up a view (similar to the view in databases) of some groups
+of one RisingWave component, in a `RisingWave` object, to expose the scale subresource. Then the magic happens and
+everything that depends on the scale subresource shall work.
+
+Take the auto-scaler as an example, we can set up the HPA or KEDA for frontend Pods with
+a `RisingWaveComponentScaleView` that targets the frontend groups.
+
+> What is the expected outcome?
+
+1. A CR `RisingWaveComponentScaleView` and its controller, with the following features:
+    - Users can target any of the four components in it, but aren't able to change it after.
+    - Users can target one or more valid groups in it, but aren't able to change it after.
+    - Users can set the priorities of groups for scaling out and in separately.
+    - It reflects all the fields required by the scale subresource described above and keeps them synced with the
+      RisingWave's.
+    - It locks the target RisingWave object on the specified groups and prevents the direct modifications (either update
+      or delete) on the related fields.
+        - It acquires the lock at the creation and removes it before deletion.
+
+2. An improvement on `RisingWave` resource and validating webhook to support lock of groups:
+    - Lock fields in the status, which is structured and easy to lookup.
+    - Enhanced validating webhook to reject any changes from besides the controller of `RisingWaveComponentScaleView`.
+
+# **Explanation**
+
+> This is the technical portion of the RFC. Explain the design in sufficient detail.
+
+## `RisingWaveComponentScaleView` CRD
+
+```yaml
+apiVersion: risingwave.risingwavelabs.com/v1alpha1
+kind: RisingWaveComponentScaleView
+metadata:
+  name: hello-frontend-scale-view
+spec:
+  targetRef:
+    name: hello             # Name of the RisingWave object, in the same namespace.
+    component: frontend     # One of meta, frontend, compute, and compactor.
+  replicas: 10              # Target replicas. It will be synced from the RisingWave object when the controller grabs the lock.
+  # Strict verification mode on replicas:
+  #   1. If enabled, the validating webhook will reject any invalid changes on .spec.replicas (e.g., exceeding the min/max range)
+  #   2. If disabled, the mutating webhook will adjust the .spec.replicas to be enclosed by the range.
+  # Serialized label selector. Would be set by the webhook.
+  strict: false
+  labelSelector: "risingwave/name=hello,risingwave/component=frontend,risingwave/group in (enduring,spot)"
+  scalePolicy: # An array of groups and the policies for scale, 
```

Comment:
how about adding or moving the `UpgradeStrategy` into the `scalePolicy`?

Reply:
It makes sense to have an upgrade strategy here and sync it to the RisingWave. But I think it's a bit complicated if we add it here. Moving won't be an option because the scale view is an optional resource and should not affect the original functionalities. Since the lock described doesn't forbid the updates on the fields besides `replicas`, I guess we can omit the upgrade strategy and let the policy focus on the distribution of replicas among groups.

Reply:
I Agree!